### PR TITLE
Modify API for advanced settings (LightGBM)

### PR DESCRIPTION
**Pull request by abgoswam**
Towards #1798 .

This PR addresses the following algos 

- LightGbmRegressorTrainer
- LightGbmBinaryTrainer
- LightGbmRankingTrainer
- LightGbmMulticlassTrainer

The following changes have been made:

1.  Two `public` extension methods, one for simple arguments and the other for advanced options
2. Make constructors `internal` .  Also a few other fields have been made `internal`
3. Pass Options objects as arguments instead of Action delegate. Also, added  3 fields to `Options`  for API consistency. 
4. Rename Arguments to Options
5. Rename Options objects as options (instead of args or advancedSettings used so far)

### CODE REVIEW

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs [hunk 1]
```diff
@@ -44,17 +45,53 @@ public static class LightGbmStaticExtensions
...
+        /// <param name="weights">The weights column.</param>
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained. Note that this action cannot change the result in any way;
+        /// it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The Score output column indicating the predicted value.</returns>
+        public static Scalar<float> LightGbm(this RegressionCatalog.RegressionTrainers catalog,
+            Scalar<float> label, Vector<float> features, Scalar<float> weights,
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: `FastTree` just has `catalog` and `options` for its input parameters when there is an explicit `options` field. Is there a standard for what the "`options` API" looks like? If not, let's make one. This needs to be consistent across the learners.

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs [hunk 2]
```diff
@@ -93,17 +129,56 @@ public static class LightGbmStaticExtensions
...
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained. Note that this action cannot change the result in any way;
+        /// it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted binary classification score (which will range
+        /// from negative to positive infinity), the calibrated prediction (from 0 to 1), and the predicted label.</returns>
+        public static (Scalar<float> score, Scalar<float> probability, Scalar<bool> predictedLabel) LightGbm(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,
+            Scalar<bool> label, Vector<float> features, Scalar<float> weights,
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: Ditto "`options` API" comment.

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs [hunk 3]
```diff
@@ -139,18 +213,59 @@ public static class LightGbmStaticExtensions
...
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained. Note that this action cannot change the result in any way;
+        /// it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted binary classification score (which will range
+        /// from negative to positive infinity), the calibrated prediction (from 0 to 1), and the predicted label.</returns>
+        public static Scalar<float> LightGbm<TVal>(this RankingCatalog.RankingTrainers catalog,
+           Scalar<float> label, Vector<float> features, Key<uint, TVal> groupId, Scalar<float> weights,
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: Ditto "`options` API" comment.

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs [hunk 4]
```diff
@@ -196,17 +307,57 @@ public static (Vector<float> score, Key<uint, TVal> predictedLabel)
...
+        /// result in any way; it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted per-class likelihoods (between 0 and 1, and summing up to 1), and the predicted label.</returns>
+        public static (Vector<float> score, Key<uint, TVal> predictedLabel)
+            LightGbm<TVal>(this MulticlassClassificationCatalog.MulticlassClassificationTrainers catalog,
+            Key<uint, TVal> label,
+            Vector<float> features,
+            Scalar<float> weights,
+            Options options,
+            Action<OvaModelParameters> onFit = null)
+        {
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: Ditto "`options` API" comment.

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs [hunk 5]
```diff
@@ -196,17 +307,57 @@ public static (Vector<float> score, Key<uint, TVal> predictedLabel)
...
+        /// the linear model that was trained. Note that this action cannot change the
+        /// result in any way; it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted per-class likelihoods (between 0 and 1, and summing up to 1), and the predicted label.</returns>
+        public static (Vector<float> score, Key<uint, TVal> predictedLabel)
+            LightGbm<TVal>(this MulticlassClassificationCatalog.MulticlassClassificationTrainers catalog,
+            Key<uint, TVal> label,
+            Vector<float> features,
+            Scalar<float> weights,
+            Options options,
+            Action<OvaModelParameters> onFit = null)
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: Why `OVA`? Is this just one-versus-all under the hood?

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 6]
```diff
@@ -98,13 +109,24 @@ public static class LightGbmExtensions
...
             int? minDataPerLeaf = null,
             double? learningRate = null,
-            int numBoostRound = LightGbmArguments.Defaults.NumBoostRound,
-            Action<LightGbmArguments> advancedSettings = null)
+            int numBoostRound = Options.Defaults.NumBoostRound)
         {
             Contracts.CheckValue(catalog, nameof(catalog));
             var env = CatalogUtils.GetEnvironment(catalog);
-            return new LightGbmRankingTrainer(env, labelColumn, featureColumn, groupIdColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound, advancedSettings);
+            return new LightGbmRankingTrainer(env, labelColumn, featureColumn, groupIdColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound);
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: There are some pretty long lines in this file. Over the permitted length?

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 7]
```diff
@@ -98,13 +109,24 @@ public static class LightGbmExtensions
...
+        }
 
+        /// <summary>
+        /// Predict a target using a decision tree binary classification model trained with the <see cref="LightGbmRankingTrainer"/>.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: binary classification => ranking

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 8]
```diff
@@ -118,24 +140,31 @@ public static class LightGbmExtensions
...
+            int numBoostRound = Options.Defaults.NumBoostRound)
         {
             Contracts.CheckValue(catalog, nameof(catalog));
             var env = CatalogUtils.GetEnvironment(catalog);
-            return new LightGbmMulticlassTrainer(env, labelColumn, featureColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound, advancedSettings);
+            return new LightGbmMulticlassTrainer(env, labelColumn, featureColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound);
+        }
 
+        /// <summary>
+        /// Predict a target using a decision tree binary classification model trained with the <see cref="LightGbmRankingTrainer"/>.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: binary classification => ranking

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 9]
```diff
@@ -54,24 +62,31 @@ public static class LightGbmExtensions
...
+        /// <param name="options">Advanced options to the algorithm.</param>
+        public static LightGbmBinaryTrainer LightGbm(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,
+            Options options)
+        {
+            Contracts.CheckValue(catalog, nameof(catalog));
+            var env = CatalogUtils.GetEnvironment(catalog);
+            return new LightGbmBinaryTrainer(env, options);
         }
 
         /// <summary>
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: binary classification => ranking

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 10]
```diff
@@ -781,7 +781,7 @@ public void TestMultiClassEnsembleCombiner()
 
             var predictors = new PredictorModel[]
             {
-                LightGbm.TrainMultiClass(Env, new LightGbmArguments
+                LightGbm.TrainMultiClass(Env, new Options
```
Comment (rogancarr) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Do you have tests for the catalog entries? I don't see them here &mdash; they didn't show up as a change, so it looks like they aren't being tested.

File: test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs [hunk 11]
```diff
@@ -216,8 +221,9 @@ public void TweedieRegressorEstimator()
         public void FastForestRegressorEstimator()
         {
             var dataView = GetRegressionPipeline();
-            var trainer = ML.Regression.Trainers.FastForest( 
-                new FastForestRegression.Options { 
+            var trainer = ML.Regression.Trainers.FastForest(
+                new FastForestRegression.Options
+                {
```
Comment (rogancarr) [inline test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs]: nit: Usually I'd pack cleanup for a different section in a different commit / PR.

Comment (abgoswam) [reply test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs]: makes sense.. i think i did a control + k + d  on this file  which did some of the cleanups

---
In reply to: [251538653](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251538653) [](ancestors = 251538653)

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs (hunk 1 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: These are the `static` extensions.   

---
In reply to: [251533791](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251533791) [](ancestors = 251533791)

File: src/Microsoft.ML.LightGBM/LightGbmBinaryTrainer.cs [hunk 12]
```diff
@@ -111,20 +111,15 @@ internal LightGbmBinaryTrainer(IHostEnvironment env, LightGbmArguments args)
         /// <param name="numBoostRound">Number of iterations.</param>
         /// <param name="minDataPerLeaf">The minimal number of documents allowed in a leaf of the tree, out of the subsampled data.</param>
         /// <param name="learningRate">The learning rate.</param>
-        /// <param name="advancedSettings">A delegate to set more settings.
-        /// The settings here will override the ones provided in the direct signature,
-        /// if both are present and have different values.
-        /// The columns names, however need to be provided directly, not through the <paramref name="advancedSettings"/>.</param>
-        public LightGbmBinaryTrainer(IHostEnvironment env,
+        internal LightGbmBinaryTrainer(IHostEnvironment env,
             string labelColumn = DefaultColumnNames.Label,
```
Comment (artidoro) [inline src/Microsoft.ML.LightGBM/LightGbmBinaryTrainer.cs]: Is it not possible to combine the two constructors?
Or have you added this work to the reconciliation issue #2100?

File: src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs [hunk 13]
```diff
@@ -51,20 +51,15 @@ internal LightGbmMulticlassTrainer(IHostEnvironment env, LightGbmArguments args)
         /// <param name="numBoostRound">Number of iterations.</param>
         /// <param name="minDataPerLeaf">The minimal number of documents allowed in a leaf of the tree, out of the subsampled data.</param>
         /// <param name="learningRate">The learning rate.</param>
-        /// <param name="advancedSettings">A delegate to set more settings.
-        /// The settings here will override the ones provided in the direct signature,
-        /// if both are present and have different values.
-        /// The columns names, however need to be provided directly, not through the <paramref name="advancedSettings"/>.</param>
-        public LightGbmMulticlassTrainer(IHostEnvironment env,
+        internal LightGbmMulticlassTrainer(IHostEnvironment env,
             string labelColumn = DefaultColumnNames.Label,
```
Comment (artidoro) [inline src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs]: Same question here, regarding unifying the two constructors?

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs (hunk 2 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: these are static extensions  (details in other comment)

---
In reply to: [251533977](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251533977) [](ancestors = 251533977)

File: src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs [hunk 14]
```diff
@@ -98,21 +98,16 @@ internal LightGbmRankingTrainer(IHostEnvironment env, LightGbmArguments args)
         /// <param name="numBoostRound">Number of iterations.</param>
         /// <param name="minDataPerLeaf">The minimal number of documents allowed in a leaf of the tree, out of the subsampled data.</param>
         /// <param name="learningRate">The learning rate.</param>
-        /// <param name="advancedSettings">A delegate to set more settings.
-        /// The settings here will override the ones provided in the direct signature,
-        /// if both are present and have different values.
-        /// The columns names, however need to be provided directly, not through the <paramref name="advancedSettings"/>.</param>
-        public LightGbmRankingTrainer(IHostEnvironment env,
+        internal LightGbmRankingTrainer(IHostEnvironment env,
             string labelColumn = DefaultColumnNames.Label,
```
Comment (artidoro) [inline src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs]: Here as well?

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs (hunk 3 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: these are static extensions  (details in other comment)

---
In reply to: [251534206](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251534206) [](ancestors = 251534206)

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs (hunk 4 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: these are static extensions  (details in other comment)

---
In reply to: [251534372](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251534372) [](ancestors = 251534372)

File: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs (hunk 5 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs]: For multiclass,  LightGBM uses the OvaModelParameters

We can also see it in the class definition as well

[https://github.com/dotnet/machinelearning/blob/620ca893a2545ed57f65c324c9c26e171ffc4839/src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs#L25-L27](https://github.com/dotnet/machinelearning/blob/620ca893a2545ed57f65c324c9c26e171ffc4839/src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs#L25-L27)


---
In reply to: [251534510](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251534510) [](ancestors = 251534510)

File: src/Microsoft.ML.LightGBM/LightGbmRegressionTrainer.cs [hunk 15]
```diff
@@ -96,25 +96,20 @@ public sealed class LightGbmRegressorTrainer : LightGbmTrainerBase<float, Regres
         /// <param name="numBoostRound">Number of iterations.</param>
         /// <param name="minDataPerLeaf">The minimal number of documents allowed in a leaf of the tree, out of the subsampled data.</param>
         /// <param name="learningRate">The learning rate.</param>
-        /// <param name="advancedSettings">A delegate to set more settings.
-        /// The settings here will override the ones provided in the direct signature,
-        /// if both are present and have different values.
-        /// The columns names, however need to be provided directly, not through the <paramref name="advancedSettings"/>.</param>
-        public LightGbmRegressorTrainer(IHostEnvironment env,
+        internal LightGbmRegressorTrainer(IHostEnvironment env,
             string labelColumn = DefaultColumnNames.Label,
```
Comment (artidoro) [inline src/Microsoft.ML.LightGBM/LightGbmRegressionTrainer.cs]: Here as well?

File: src/Microsoft.ML.LightGBM/LightGbmBinaryTrainer.cs (hunk 12 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmBinaryTrainer.cs]: yeap. exactly.  added to #2100

want to focus on cleanup of public surface first

---
In reply to: [251541952](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251541952) [](ancestors = 251541952)

File: src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs (hunk 13 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs]: #2100

---
In reply to: [251542646](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251542646) [](ancestors = 251542646)

File: src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs (hunk 14 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs]: #2100

---
In reply to: [251542949](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251542949) [](ancestors = 251542949)

File: src/Microsoft.ML.LightGBM/LightGbmRegressionTrainer.cs (hunk 15 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmRegressionTrainer.cs]: #2100

---
In reply to: [251545552](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251545552) [](ancestors = 251545552)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 8 above)
Comment (artidoro) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: (MultiClass)

---
In reply to: [251536106](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251536106) [](ancestors = 251536106)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 16]
```diff
@@ -118,24 +140,31 @@ public static class LightGbmExtensions
...
             Contracts.CheckValue(catalog, nameof(catalog));
             var env = CatalogUtils.GetEnvironment(catalog);
-            return new LightGbmMulticlassTrainer(env, labelColumn, featureColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound, advancedSettings);
+            return new LightGbmMulticlassTrainer(env, labelColumn, featureColumn, weights, numLeaves, minDataPerLeaf, learningRate, numBoostRound);
+        }
 
+        /// <summary>
+        /// Predict a target using a decision tree binary classification model trained with the <see cref="LightGbmRankingTrainer"/>.
+        /// </summary>
+        /// <param name="catalog">The <see cref="RankingCatalog"/>.</param>
```
Comment (artidoro) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: >RankingCatalog [](start = 49, length = 14)

Should be MultiClass catalog I think.

**Comment by artidoro** [comment]
>         /// <param name="catalog">The <see cref="RankingCatalog"/>.</param>

I know you haven't written this section, but it should also say MultiClass instead of ranking.

---
Refers to: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs:135 in 33fe723. [](commit_id = 33fe7235393c1b277f82a57a32a646fb43281773, deletion_comment = False)

**Comment by artidoro** [comment]
>         /// <param name="catalog">The <see cref="RankingCatalog"/>.</param>

It would be great if you could fix it as part of the PR, as it's related.

---
In reply to: [458257429](https://github.com/dotnet/machinelearning/pull/2261#issuecomment-458257429) [](ancestors = 458257429)

---
Refers to: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs:135 in 33fe723. [](commit_id = 33fe7235393c1b277f82a57a32a646fb43281773, deletion_comment = False)

**Comment by abgoswam** [comment]
>         /// <param name="catalog">The <see cref="RankingCatalog"/>.</param>

fixed

---
In reply to: [458258712](https://github.com/dotnet/machinelearning/pull/2261#issuecomment-458258712) [](ancestors = 458258712,458257429)

---
Refers to: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs:135 in 33fe723. [](commit_id = 33fe7235393c1b277f82a57a32a646fb43281773, deletion_comment = False)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 16 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: fixed

---
In reply to: [251548528](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251548528) [](ancestors = 251548528)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 6 above)
Comment (abgoswam) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: i do not see any warnings related to long lines. 

i am actually reducing the length, by dropping `advancedSettings` from the line :)

---
In reply to: [251535698](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251535698) [](ancestors = 251535698)

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 10 above)
Comment (abgoswam) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: good catch. I have updated several tests so they exercise the public API catalog extensions 

---
In reply to: [251537729](https://github.com/dotnet/machinelearning/pull/2261#discussion_r251537729) [](ancestors = 251537729)

**Pull request merged**